### PR TITLE
Do not trap on pruning non-existing aggregations or zero weighted events

### PR DESCRIPTION
**Pull request by frederikrothenberger**
This PR fixes a bug introduced in #2449, which subsequently led to a rollback in proposal 130068.

This refactors the code to accommodate violated expectations and proceed without trapping.
Zero weighted event are now correctly ignored by the pruning logic.

### CODE REVIEW

**Review by lmuntaner — APPROVED** [review]
Rubber stamp, LGTM!

File: src/internet_identity/src/stats/event_stats.rs [hunk 1]
```diff
@@ -420,21 +422,18 @@ fn update_aggregation<M: Memory, F>(
...
+                let new_weight = current_weight.saturating_sub(pruned_weight);
+                if new_weight == 0 {
+                    db.remove(&key);
+                } else {
+                    db.insert(key, new_weight);
+                }
             } else {
-                db.insert(key, new_weight);
+                // This should not happen, but if it does, it's a bug.
+                // Print a message and so that we can investigate once canister logs are available.
```
Comment (przydatek) [inline src/internet_identity/src/stats/event_stats.rs:435]: nit: spurious "and" after "message"

File: src/internet_identity/src/stats/event_stats.rs [hunk 2]
```diff
@@ -389,6 +389,8 @@ fn update_aggregation<M: Memory, F>(
     let mut pruned_weight_by_key = pruned_events
         .iter()
         .filter_map(&aggregation_filter_map)
+        // zero weighted events are ignored for pruning
```
Comment (przydatek) [inline src/internet_identity/src/stats/event_stats.rs:392]: nit: this comment is somewhat confusing when compared to the test `should_prune_zero_weighted_events` added below.  Maybe reformulate/add more info?

**Review by przydatek — APPROVED** [review]
LGTM, with minor suggestions. Thanks!

File: src/internet_identity/src/stats/event_stats.rs (hunk 2 above)
Comment (frederikrothenberger) [reply src/internet_identity/src/stats/event_stats.rs]: I adjusted it. Is it clearer now?

File: src/internet_identity/src/stats/event_stats.rs [hunk 3]
```diff
@@ -389,6 +389,10 @@ fn update_aggregation<M: Memory, F>(
     let mut pruned_weight_by_key = pruned_events
         .iter()
         .filter_map(&aggregation_filter_map)
+        // zero weighted events are ignored when adjusting aggregations during pruning because they
+        // don't contribute any change to the aggregations hence pruning them has no effect on
+        // aggregations
+        .filter(|weighted_aggregation| weighted_aggregation.weight > 0)
```
Comment (lmuntaner) [inline src/internet_identity/src/stats/event_stats.rs:395]: If you remove this, does the test still pass?

File: src/internet_identity/src/stats/event_stats/tests.rs [hunk 4]
```diff
@@ -637,6 +637,60 @@ fn should_add_prune_events_to_prune_aggregations() {
...
+            session_duration_ns: to_ns(0),
+        }),
+    };
+
+    // no events initially
+    assert_eq!(storage.event_data.len(), 0);
+
+    // add multiple zero weighted event
+    update_events_internal(event.clone(), TIMESTAMP, &mut storage);
+    update_events_internal(event.clone(), TIMESTAMP, &mut storage);
```
Comment (lmuntaner) [inline src/internet_identity/src/stats/event_stats/tests.rs:657]: Does it make sense to add a zero event from a different frontend (and therefore a different key)?

**Review by lmuntaner — COMMENTED** [review]
A couple of comments, but it looks good to me!

File: src/internet_identity/src/stats/event_stats.rs (hunk 3 above)
Comment (frederikrothenberger) [reply src/internet_identity/src/stats/event_stats.rs:395]: Yes.

File: src/internet_identity/src/stats/event_stats/tests.rs (hunk 4 above)
Comment (frederikrothenberger) [reply src/internet_identity/src/stats/event_stats/tests.rs:657]: Yes, done.

File: src/internet_identity/src/stats/event_stats.rs [hunk 5]
```diff
@@ -420,21 +424,18 @@ fn update_aggregation<M: Memory, F>(
...
+                if new_weight == 0 {
+                    db.remove(&key);
+                } else {
+                    db.insert(key, new_weight);
+                }
             } else {
-                db.insert(key, new_weight);
+                // This should not happen, but if it does, it's a bug.
+                // Print a message so that we can investigate once canister logs are available.
+                ic_cdk::println!("WARN: Aggregation key {:?} not found in db", key);
```
Comment (cybrowl) [inline src/internet_identity/src/stats/event_stats.rs:438]: Is this `key` a public key? @frederikrothenberger

Comment (frederikrothenberger) [reply src/internet_identity/src/stats/event_stats.rs:438]: @cybrowl: No, this is a key into a map of event aggregations that we use to collect usage statistics of II. See [here](https://github.com/dfinity/internet-identity/blob/8d2edeb81d59e77ac85d4421a48c22c09d9db8c7/src/internet_identity/src/stats/event_stats/event_aggregations.rs#L256).

**Pull request merged**